src/lib/helpers/decorators.tsx

```diff
@@ -0,0 +1,53 @@
+import * as Analytics from "../../analytics/analytics"
+
+// This heavily uses ideas and approaches from @autobind
+// https://github.com/andreypopp/autobind-decorator
+export default function eventHandler(analyticsConstant: string) {
+  // TODO: disable tslint arrow function rule
+  return function bindMethod(target: any, propertyKey: string, descriptor: PropertyDescriptor) {
+    let originalMethod = descriptor.value
+
+    if (typeof originalMethod !== "function") {
+      throw new Error(`@eventHandler can only be applied to methods not: ${typeof originalMethod}`)
+    }
+
+    // normally in decorators, you overwrite the descriptor's value (= function), and return as is
+    // because we're binding, we're creating a whole new descriptor
+    let newDescriptor = {
+      configurable: true,
+      get() {
+        if (this === target.prototype || this.hasOwnProperty(propertyKey)) {
+          return originalMethod
+        }
+
+        let boundFn = originalMethod.bind(this)
+
+        let analyticsWrappedMethod
+        if (analyticsConstant) {
+          let contextualProps = this.props
+          Analytics.sendIntent(analyticsConstant, contextualProps)
+          analyticsWrappedMethod = function (...args: any[]) {
+            const resultPromise = boundFn.apply(this, args)
+            resultPromise
```

Comment:
As mentioned above, you can do

`Promise.resolve(boundFn.apply(this, args)).then(...)` instead, it'll safely ensure that `.then` gets called regardless of the return type of your boundFunction